CHANGELOG.md

```diff
@@ -45,6 +45,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 ### Fixed
 
 - Flaky decimal/TestSelect (#300)
+- Incorrect decoding of an MP_DECIMAL when the `scale` value is negative (#314)
```

Comment:
Is it really valid? It doesn't seem that Tarantool works fine with negative scale itself

https://github.com/tarantool/tarantool/blob/ba749e820bf0638aa3f79f266848590f9713c1cf/src/lib/core/decimal.c#L289-L231

Reply:
> Yes, it is:
> 
> https://github.com/tarantool/tarantool/blob/ba749e820bf0638aa3f79f266848590f9713c1cf/src/lib/core/decimal.c#L503-L511

And at the same time
https://github.com/tarantool/tarantool/blob/ba749e820bf0638aa3f79f266848590f9713c1cf/src/lib/core/decimal.c#L102

Reply:
It is for:

https://www.tarantool.io/ru/doc/latest/reference/reference_lua/decimal/#lua-function.decimal.scale

Reply:
```
> conn = net_box:connect('localhost:3013')
---
...
tarantool> conn:eval("return require('decimal').new('1e2')")
---
- 100
...
```

Reply:
> ```
> > conn = net_box:connect('localhost:3013')
> ---
> ...
> tarantool> conn:eval("return require('decimal').new('1e2')")
> ---
> - 100
> ...
> ```

Well, this example doesn't say anything about the scale -- underneath it could be "digits 1 0 0 with scale 0" (as I thought they will always be).

Reply:
```
tarantool> msgpack.encode(conn:eval("return require('decimal').new('1')")):gsub('.', function(c) return string.format('%02x', string.byte(c)) end)
---
- d501001c
- 4
...

tarantool> msgpack.encode(conn:eval("return require('decimal').new('1e2')")):gsub('.', function(c) return string.format('%02x', string.byte(c)) end)
---
- d501fe1c
- 4
...

tarantool> msgpack.encode(conn:eval("return require('decimal').new('100')")):gsub('.', function(c) return string.format('%02x', string.byte(c)) end)
---
- c7030100100c
- 6
...

```

See `1` vs `1e2` difference:

`d501_00_1c` vs `d501_fe_1c`

Reply:
`0xd5`:

https://github.com/vmihailenco/msgpack/blob/v5/msgpcode/msgpcode.go#L49

`0x01`:

https://github.com/tarantool/go-tarantool/blob/445c3aef5526c58ff5ed15dc3a40594ad9202bb1/decimal/decimal.go#L36

`0x00` and `0xfe` are scale values.

Reply:
> Btw, you can use `str:hex()` for that. Yeah, I have seen msgpack representation for such numbers, but it's not clear for me from the source code and documentation whether it is expected or unexpected. I have asked @sergepetrenko to provide some comments as a core package developer.

Yes, it is not obvious from the documentation, but from the source code we expect that `scale` may be < 0:

https://github.com/tarantool/tarantool/blob/ba749e820bf0638aa3f79f266848590f9713c1cf/src/lib/core/decimal.c#L432-L450
+
`decPackedFromNumber`

https://github.com/tarantool/tarantool/blob/ba749e820bf0638aa3f79f266848590f9713c1cf/src/lib/core/decimal.c#L479-L533
+
`decPackedToNumber`

Reply:
Thank you for the detailed explanation

Reply:
@oleg-jukovec is right, we encode scale as negated exponent. So, for numbers like 1e15, scale will be `-15`.
And it will be positive for numbers with nonzero fractional part, like `164.23` here the scale is `2`.

So the semantics of `scale` is "by how many digits you have to move the decimal point left or right to get the correct number".